### PR TITLE
Cut over all index scope settings to the new setting infrastrucuture

### LINKED ISSUE #6732: Improve cluster/index settings
Currently there is no way of seeing all settings that are enforced.  For instance, defaults are not available and settings from the `config/elasticsearch.yml` are not available.

Any setting which is different from the default should be returned by these APIs:

    GET /{index}/_settings
    GET /_cluster/settings

Both should also support the `include_defaults` query string parameter, which would return all settings, including the defaults.

Also, deleting a setting should reset the setting to the default.

Setting an unknown setting, or a setting that can't be changed should throw an error.

Relates to #5018, #3670, #2738, #2730, #3572, #2628, #3671, #5019, #6309

### LINKED ISSUE #12790: Index "creation_date" not accurate when created with settings from another index
When creating a new index using settings from an old index, the "creation_date" of new index is same as that of old index. Steps - 
1. Get settings of existing index using GetIndexRequestBuilder()
2. Create new index using CreateIndexRequestBuilder() by doing - 
a. copying the mappings getIndexResponse.getMappings()
b. copying the settings getIndexResponse.getSettings()
c. copying headers getIndexResponse.getHeaders()

Now the new index has same "creation_date" timestamp as old index. Version - 1.4

### LINKED ISSUE #16032: Setting invalid codec causes shard to keep failing and stuck in initializing state
On ES 2.1.1.  Ran into this when trying to verify if a setting has persisted.  In this case, I am purposely setting the codec to an invalid codec name `best_compression1` instead of `best_compression`:

```
POST /largeevent/_close
PUT /largeevent/_settings
{
  "index.codec":"best_compression1"
}
POST /largeevent/_open
GET /largeevent/_settings
```

As soon as, I reopen the index after making the setting change, the shard gets into a failure loop trying to initialize the shard and will not start up.

```
[2016-01-15 18:12:18,597][WARN ][cluster.action.shard     ] [node-1] [largeevent][1] received shard failed for [largeevent][1], node[O373snrkTTGclRFL0t7H5g], [P], v[1553], s[INITIALIZING], a[id=w3VoZACfQKq_QcsDGLaZ3Q], unassigned_info[[reason=ALLOCATION_FAILED], at[2016-01-16T02:12:18.563Z], details[failed recovery, failure IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException[failed to find codec [best_compression1]]; ]], indexUUID [JEPoX4QkR_Om-W69VKV6iQ], message [failed recovery], failure [IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException[failed to find codec [best_compression1]]; ]
[largeevent][[largeevent][1]] IndexShardRecoveryException[failed recovery]; nested: IllegalArgumentException[failed to find codec [best_compression1]];
	at org.elasticsearch.index.shard.StoreRecoveryService$1.run(StoreRecoveryService.java:179)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: failed to find codec [best_compression1]
	at org.elasticsearch.index.codec.CodecService.codec(CodecService.java:94)
	at org.elasticsearch.index.engine.EngineConfig.getCodec(EngineConfig.java:255)
	at org.elasticsearch.index.engine.InternalEngine.createWriter(InternalEngine.java:1057)
	at org.elasticsearch.index.engine.InternalEngine.<init>(InternalEngine.java:150)
	at org.elasticsearch.index.engine.InternalEngineFactory.newReadWriteEngine(InternalEngineFactory.java:25)
	at org.elasticsearch.index.shard.IndexShard.newEngine(IndexShard.java:1408)
	at org.elasticsearch.index.shard.IndexShard.createNewEngine(IndexShard.java:1403)
	at org.elasticsearch.index.shard.IndexShard.internalPerformTranslogRecovery(IndexShard.java:906)
	at org.elasticsearch.index.shard.IndexShard.performTranslogRecovery(IndexShard.java:883)
	at org.elasticsearch.index.shard.StoreRecoveryService.recoverFromStore(StoreRecoveryService.java:245)
	at org.elasticsearch.index.shard.StoreRecoveryService.access$100(StoreRecoveryService.java:56)
	at org.elasticsearch.index.shard.StoreRecoveryService$1.run(StoreRecoveryService.java:129)
	... 3 more
```

Will be nice if we can handle this better.

### PR DESCRIPTION
**Pull request by s1monw**
This change moves all `index.*` settings over to the new infrastructure. This means in short that:
- every setting that has an index scope must be registered up-front
- index settings are validated on index creation,  template creation, index settings update
- node level settings starting with `index.*` are validated on node startup
- settings that are private to ES like `index.version.created` can only be set by tests when they install a specific test plugin.
- all index settings can be reset by passing `null` as their value on update
- all index settings defaults can be listed via the settings APIs

Closes #12854
Closes #6732
Closes #16032
Closes #12790

Thanks @brwe who helped me a lot by going through the mechanical part of converting settings. Much appreciated!

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/settings/IndexScopeSettings.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import org.elasticsearch.search.SearchService;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * Encapsulates all valid cluster level settings.
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/IndexScopeSettings.java]: Wrong javadoc?

File: core/src/main/java/org/elasticsearch/common/settings/IndexScopeSettings.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+ */
+public final class IndexScopeSettings extends AbstractScopedSettings {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/IndexScopeSettings.java]: s/IndexScopeSettings/IndexScopedSettings/ ?

File: core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ */
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.common.Booleans;
+import org.elasticsearch.common.settings.Setting;
+
+/**
+ * This class acts as a functional wrapper around the <tt>index.auto_expand_replicas</tt> setting.
+ * This setting or rather it's value is expanded into a min and max value which requires special handling
+ * based on the number of datanodes in the cluster. This class handels all the parsing and streamlines the access to these values.
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java]: s/handels/handles/

Handel was a composer.

File: core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return minReplicas;
+    }
+
+    int getMaxReplicas(int numDataNodes) {
+        return Math.min(maxReplicas, numDataNodes-1);
+    }
+
+    @Override
+    public String toString() {
+        return enabled == false ? Boolean.toString(enabled) : minReplicas + "-" + maxReplicas;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java]: I think this'd be more clear: `enabled ? minReplicas + "-" + maxReplicas : "false"`

File: core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            }
+        }
+        return new AutoExpandReplicas(min, max, true);
+    }, true, Setting.Scope.INDEX);
+
+    private final int minReplicas;
+    private final int maxReplicas;
+    private final boolean enabled;
+
+    AutoExpandReplicas(int minReplicas, int maxReplicas, boolean enabled) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java]: Can/should this be private?

File: core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+import org.elasticsearch.common.Booleans;
+import org.elasticsearch.common.settings.Setting;
+
+/**
+ * This class acts as a functional wrapper around the <tt>index.auto_expand_replicas</tt> setting.
+ * This setting or rather it's value is expanded into a min and max value which requires special handling
+ * based on the number of datanodes in the cluster. This class handels all the parsing and streamlines the access to these values.
+ */
+final class AutoExpandReplicas {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java]: ++ on doing this

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java [hunk 7]
```diff
@@ -460,16 +464,17 @@ public void validateIndexSettings(String indexName, Settings settings) throws In
...
+        if (Strings.isEmpty(customPath) == false && env.sharedDataFile() == null) {
             validationErrors.add("path.shared_data must be set in order to use custom data paths");
-        } else if (customPath != null) {
+        } else if (Strings.isEmpty(customPath) == false) {
             Path resolvedPath = PathUtils.get(new Path[]{env.sharedDataFile()}, customPath);
             if (resolvedPath == null) {
                 validationErrors.add("custom path [" + customPath + "] is not a sub-path of path.shared_data [" + env.sharedDataFile() + "]");
             }
         }
+        //norelease - this can be removed?
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java]: Not yet! Those are still just strings but I suspect this note is to make sure that they grow their validation at the setting level.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java [hunk 8]
```diff
@@ -150,7 +149,7 @@ private static boolean isSupportedVersion(IndexMetaData indexMetaData) {
                                     "index.translog.interval",
                                     "index.translog.sync_interval",
                                     "index.shard.inactive_time",
-                                    UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING));
+                                    "index.unassigned.node_left.delayed_timeout"));
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: Should these be settings references? I'm not sure why they aren't but I'm sure there was a good reason but I don't know if its still true.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java [hunk 9]
```diff
@@ -90,64 +86,30 @@ public void clusterChanged(ClusterChangedEvent event) {
...
-                            logger.warn("failed to set [{}] for index [{}], maximum value is neither [{}] nor a number [{}]",
-                                    e, IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS, indexMetaData.getIndex(), ALL_NODES_VALUE, sMax);
-                            continue;
-                        }
-                    }
+            AutoExpandReplicas autoExpandReplicas = IndexMetaData.SETTING_AUTO_EXPAND_REPLICAS_SETTING.get(indexMetaData.getSettings());
+            if (autoExpandReplicas.isEnabled()) {
+                final int min = autoExpandReplicas.getMinReplicas();
+                final int max = autoExpandReplicas.getMaxReplicas(dataNodeCount);
+                int numberOfReplicas = dataNodeCount - 1;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java]: I know you are just keeping this logic from before but it'd be cool to have some comment explaining it.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 10]
```diff
@@ -152,14 +152,29 @@ public static State fromString(String state) {
...
+    public static final Setting<Integer> INDEX_NUMBER_OF_REPLICAS_SETTING = Setting.intSetting(SETTING_NUMBER_OF_REPLICAS, 1, 0, true, Setting.Scope.INDEX);
     public static final String SETTING_SHADOW_REPLICAS = "index.shadow_replicas";
+    public static final Setting<Boolean> INDEX_SHADOW_REPLICAS_SETTING = Setting.boolSetting(SETTING_SHADOW_REPLICAS, false, false, Setting.Scope.INDEX);
+
     public static final String SETTING_SHARED_FILESYSTEM = "index.shared_filesystem";
+    public static final Setting<Boolean> INDEX_SHARED_FILESYSTEM_SETTING = Setting.boolSetting(SETTING_SHARED_FILESYSTEM, false, false, Setting.Scope.INDEX);
+
     public static final String SETTING_AUTO_EXPAND_REPLICAS = "index.auto_expand_replicas";
+    public static final Setting<AutoExpandReplicas> SETTING_AUTO_EXPAND_REPLICAS_SETTING = AutoExpandReplicas.SETTING;
     public static final String SETTING_READ_ONLY = "index.blocks.read_only";
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Having both the SETTING_$NAME and INDEX_$NAME_SETTING is a bit confusing.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 11]
```diff
@@ -152,14 +152,29 @@ public static State fromString(String state) {
...
     public static final String SETTING_NUMBER_OF_REPLICAS = "index.number_of_replicas";
+    public static final Setting<Integer> INDEX_NUMBER_OF_REPLICAS_SETTING = Setting.intSetting(SETTING_NUMBER_OF_REPLICAS, 1, 0, true, Setting.Scope.INDEX);
     public static final String SETTING_SHADOW_REPLICAS = "index.shadow_replicas";
+    public static final Setting<Boolean> INDEX_SHADOW_REPLICAS_SETTING = Setting.boolSetting(SETTING_SHADOW_REPLICAS, false, false, Setting.Scope.INDEX);
+
     public static final String SETTING_SHARED_FILESYSTEM = "index.shared_filesystem";
+    public static final Setting<Boolean> INDEX_SHARED_FILESYSTEM_SETTING = Setting.boolSetting(SETTING_SHARED_FILESYSTEM, false, false, Setting.Scope.INDEX);
+
     public static final String SETTING_AUTO_EXPAND_REPLICAS = "index.auto_expand_replicas";
+    public static final Setting<AutoExpandReplicas> SETTING_AUTO_EXPAND_REPLICAS_SETTING = AutoExpandReplicas.SETTING;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: The name of this one seems wrong.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java [hunk 12]
```diff
@@ -267,57 +221,67 @@ public ClusterState execute(ClusterState currentState) {
...
                 ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
-                Boolean updatedReadOnly = openSettings.getAsBoolean(IndexMetaData.SETTING_READ_ONLY, null);
-                if (updatedReadOnly != null) {
-                    for (String index : actualIndices) {
-                        if (updatedReadOnly) {
-                            blocks.addIndexBlock(index, IndexMetaData.INDEX_READ_ONLY_BLOCK);
-                        } else {
-                            blocks.removeIndexBlock(index, IndexMetaData.INDEX_READ_ONLY_BLOCK);
-                        }
+                final boolean updatedReadOnly = IndexMetaData.INDEX_READ_ONLY_SETTING.get(openSettings);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataUpdateSettingsService.java]: If looks like if you didn't specify the value it used to leave it unchanged on all the indexes but now it'll remove all the blocks I think.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java [hunk 13]
```diff
@@ -61,10 +61,10 @@
     public static final String NAME = "enable";
 
     public static final Setting<Allocation> CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING = new Setting<>("cluster.routing.allocation.enable", Allocation.ALL.name(), Allocation::parse, true, Setting.Scope.CLUSTER);
-    public static final String INDEX_ROUTING_ALLOCATION_ENABLE= "index.routing.allocation.enable";
+    public static final Setting<Allocation> INDEX_ROUTING_ALLOCATION_ENABLE_SETTING = new Setting<>("index.routing.allocation.enable", Allocation.ALL.name(), Allocation::parse, true, Setting.Scope.INDEX);
 
     public static final Setting<Rebalance> CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING = new Setting<>("cluster.routing.rebalance.enable", Rebalance.ALL.name(), Rebalance::parse, true, Setting.Scope.CLUSTER);
-    public static final String INDEX_ROUTING_REBALANCE_ENABLE = "index.routing.rebalance.enable";
+    public static final Setting<Rebalance> INDEX_ROUTING_REBALANCE_ENABLE_SETTING = new Setting<>("index.routing.rebalance.enable", Rebalance.ALL.name(), Rebalance::parse, true, Setting.Scope.INDEX);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java]: Very nice

File: core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java [hunk 14]
```diff
@@ -162,6 +180,37 @@ public synchronized Settings applySettings(Settings newSettings) {
     }
 
     /**
+     * Validates that all settings in the builder are registered and valid
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java]: Its not really a builder, right?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java]: Ooops - wasn't reading right. Ignore.

File: core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java [hunk 15]
```diff
@@ -162,6 +180,37 @@ public synchronized Settings applySettings(Settings newSettings) {
...
+    /**
+     * Validates that the setting is valid
+     */
+    public final void validate(String key, String value) {
+        Settings.Builder builder = Settings.builder().put(key, value);
+        Setting setting = get(key);
+        if (setting == null) {
+            throw new IllegalArgumentException("unknown setting [" + key + "]");
+        }
+        setting.get(builder.build());
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java]: This seems like a pretty heavy way to validate a setting. Like, its building a bunch of maps and things.

File: core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java [hunk 16]
```diff
@@ -253,4 +302,66 @@ public Settings diff(Settings source, Settings defaultSettings) {
...
+        return setting.get(this.lastSettingsApplied, settings);
+    }
+
+    /**
+     * Updates a target settings builder with new, updated or deleted settings from a given settings builder.
+     * @param toApply the new settings to apply
+     * @param target the target settings builder that the updates are applied to. All keys that have explicit null value in toApply will be removed from this builder
+     * @param updates a settings builder that holds all updates applied to target
+     * @param type a free text string to allow better exceptions messages
+     * @param all if <code>true</code> all settings are updated otherwise only dynamic settings are updated. if set to <code>false</code> and a non-dynamic setting is updated an exception is thrown
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java]: maybe have `dynamicOnly` instead? It feels more precise.

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 17]
```diff
@@ -48,15 +50,30 @@
  */
 public abstract class PrimaryShardAllocator extends AbstractComponent {
 
-    @Deprecated
-    public static final String INDEX_RECOVERY_INITIAL_SHARDS = "index.recovery.initial_shards";
+    private static final Function<String, String> INITIAL_SHARDS_PARSER = (value) -> {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: It'd be nice if this got a class the same way AutoExpandReplicas did. Not required at all, but pleasant.

File: core/src/main/java/org/elasticsearch/index/IndexModule.java [hunk 18]
```diff
@@ -57,25 +59,24 @@
...
-    public static final String STORE_TYPE = "index.store.type";
+    public static final Setting<String> INDEX_STORE_TYPE_SETTING = new Setting<>("index.store.type", "", Function.identity(), false, Setting.Scope.INDEX);
     public static final String SIMILARITY_SETTINGS_PREFIX = "index.similarity";
     public static final String INDEX_QUERY_CACHE = "index";
     public static final String NONE_QUERY_CACHE = "none";
-    public static final String QUERY_CACHE_TYPE = "index.queries.cache.type";
+    public static final Setting<String> INDEX_QUERY_CACHE_TYPE_SETTING = new Setting<>("index.queries.cache.type", INDEX_QUERY_CACHE, Function.identity(), false, Setting.Scope.INDEX);
     // for test purposes only
-    public static final String QUERY_CACHE_EVERYTHING = "index.queries.cache.everything";
+    public static final Setting<Boolean> INDEX_QUERY_CACHE_EVERYTHING_SETTING = Setting.boolSetting("index.queries.cache.everything", false, false, Setting.Scope.INDEX);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/IndexModule.java]: Side note: this one's come up lately as some people have been using it as a big hammer to fix caching issues. They know its a bad idea but its the only tool they have.

**Comment by nik9000** [comment]
Its a huge improvement. I left some comments that amount to "I don't like this name" and "you just touched some undocumented code, can you document it now?" but I didn't see anything that should block merging. LGTM. If you merge I'll open another PR for some of the things I suggested.

File: core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java (hunk 8 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java]: I removed all of this - it's a pre 2.0 BWC layer we don't need anymore in master

**Pull request merged**